### ISSUE TITLE
Used chached_property

### ISSUE THREAD

**Issue by jlibovicky** [ufal/neuralmonkey#189]
What do you say to using [chached_property](https://pypi.python.org/pypi/cached-property/1.2.0) in the building blocks of the graph (encoders, decoders, attention):

* can be used efficiently separate parts of graph creating a graph without assigning to object fields
* can help us separate common functionality of encoders and decoders to abstract base classes like we did in case of attentive encoders (we can make abstract encoder, abstract sequence decoder)
* potentially allows easier trying new variants of existing parts of model by inheritance (and multiple inheritance) from existing model parts prototypes

Drawbacks:
* another dependency to our code
* another concept a potential contributor needs to get familiar with

If we decide to use it, it doesn't mean we have to immediately rewrite everything in this fashion, just have it in mind while implementing new features.

**Comment by jindrahelcl** [comment]
Could you give an example when the caching property would be of any use?

**Comment by jlibovicky** [comment]
In case of `Attentive`
```python
   ...
   self.__attention_tensor = ... # the code

@property
def _attention_tensor(self):
    return self.__attention_tensor
```
we can have
```python
@cached_property
def _attention_tensor(self):
    return ... # the code
```

Another thing we can do is create abstract base class for encoders that would have abstract property of `encoded` returning a Tensor, abstract base class for decoders having a list of logits as an abstract property a these classes could be used also for the type checking using mypy.

**Comment by jindrahelcl** [comment]
Are there any quirks of the `cached_property` decorator beside the fact it saves its first return value and then returns it? Is it exactly equivalent to this?

```
@property
def _attention_tensor(self):
    if not hasattr(self, "__attention_tensor"):
        self.__attention_tensor ... # the code
    return self.__attention_tensor
```

**Comment by jlibovicky** [comment]
Yes, you can uncache it by `del obj._attention_tensor`, there are also some ways to deal with setter for the properties. But I would use just as a syntactic sugar for what you have just written.

**Comment by jindrahelcl** [comment]
In that case, I would rather use the code I posted in the comment above, since we don't need any other properties of the `cached_property` module, such as timeout or invalidating of the cache which it provides. I think using the `cached_property` would be more rovnák na ohýbák than useful.

Maybe we could define our own decorator which would do the caching. As I undrestood it, it would be suffice to create a function `output_tensor`, e.g. in module `neuralmonkey.decorators`, as this:

```python
def output_tensor(func):
    def decorate(self, *args, **kwargs):
        attribute_name = "_{}_cached".format(func.__name__)
        if not(hasattr(self, attribute_name):
            value = func(self, *args, **kwargs)
            setattr(self, attribute_name, value)
    return decorate
```

and then

```python
@output_tensor
def _attention_tensor(self):
    return ... # the code
```

Disclaimer: there can be bugs in the code, but the idea and usage is correct, as far as I know.

There is probably a better place for the cached tensor than in the concerned object's `__dict__`.

**Comment by jlibovicky** [comment]
Our own decorator is a good idea - I didn't know it's such an easy thing.

**Comment by jindrahelcl** [comment]
Me neither, I just googled how to do it :-)

**Comment by jindrahelcl** [comment]
I'll try it in the decoder refactor

**Comment by jindrahelcl** [comment]
It seems to work. I just had to slightly alter the code:

```python
def output_tensor(func):
    # body of output_tensor
    return property(decorator)    
```
so the functions annotated by this decorator would behave like properties.

**Comment by jlibovicky** [comment]
Moved to #191.
